### ISSUE TITLE
My training goal!

### ISSUE THREAD

**Issue by oysteijo** [oysteijo/simd_neuralnet#11]
This is a goal statement. With this toolset I should be able to reach a mean squared error of the backgammon validation dataset of **1.5e-04**.

**Comment by oysteijo** [comment]
So ... what do I need to get there? I can get about 3.5e-04 today. 

I definitely need better optimizers. I may also need a regularization (L1 and L2), as I think I might be overfitting. I don't think dropout will help, but I can try it. Maybe a custom loss function can to the trick?

**Comment by oysteijo** [comment]
I can get 2.5e-4 today. It's just a matter of setting `binary_crossentropy` with `sigmoid` output activation, and that's it.

**Comment by oysteijo** [comment]
> I can get 2.5e-4 today. It's just a matter of setting `binary_crossentropy` with `sigmoid` output activation, and that's it.

Just to keep the record, I believe that was achieved with `250 144 96 48 5 relu relu relu sigmoid` I think this might have been 20 or 40 epochs of plain SGD (no momentum nor nesterov). From the code evidence it looks like I might have achieved this by having 20 epochs of lr=0.01 and another 20 epochs with lr=0.0002. See WIP on master: 093926d on the LinuxGPU4.

**Comment by oysteijo** [comment]
OK! I got mean squared error of 2.373e-4 with `250 144 96 48 5 relu relu relu sigmoid` . Using binary_crossentropy. 20 plain SGD with LR=0.05. Unfortunately I forgot to log the training. :-o

**Comment by oysteijo** [comment]
It can be squeezed further down to 2.215e-4 with extending with lower learning rates. I've now extended 20 epochs with LR=0.01 and then another 20 epochs of LR=0.002. It think that's about the limit we can get without doing further changes.

**Comment by oysteijo** [comment]
Playing around a bit. Using `adagrad` and leaving the training overnight, I can now achieve **1.94496e-04**. Using the same layer settings as above, but I use batch size of 8. It is trained for more than 600 epochs.

**Comment by oysteijo** [comment]
OK! I think I got a recipe on how to get there. I get there with Keras when I do this:

    sizes = (250, 224, 192, 160, 128, 96, 64, 32, 5)

I used `relu` all the way except the output which is `sigmoid`. Loss is `binary_crossentropy`. Adam optimizer. (Keep in mind that all kernel_initializers are just the Keras default which for Dense() layers are `glorot_uniform`, which is called `xavier` in this code.)

- Then do hundred epochs with batchsize 256,
- Another hundred epochs with batchsize 512,
- Another hundred epochs with batchsize 1024,
- Another hundred epochs with batchsize 2048,
- Another hundred epochs with batchsize 4096,
- so on

You should then reach 4.9e-05-ish train error and 1.5e-04 on the validation error. 

I'm a bit surprised that this works, I have no idea why. My experience with SGD is that smaller batch sizes works better.

**Comment by oysteijo** [comment]
> OK! I think I got a recipe on how to get there. I get there with Keras when I do this:

I think everything that is left in the C code to recreate this experiment is a actually just the Adam Optimizer. I'll put priority on that. BTW: I guess the batch code can be threaded?

**Comment by oysteijo** [comment]
I've now got Adam working with the C code, and I get these number for the above described experiment. I do add early stopping with patience 10 though.

    batch_size:  256  Epoch  46 mean_squared_error: 8.05933e-05 val_mean_squared_error: 1.63234e-04
    batch_size:  512  Epoch   1 mean_squared_error: 7.47747e-05 val_mean_squared_error: 1.56484e-04
    batch_size: 1024  Epoch  18 mean_squared_error: 7.03768e-05 val_mean_squared_error: 1.53398e-04
    batch_size: 2048  Epoch  18 mean_squared_error: 6.18713e-05 val_mean_squared_error: 1.54542e-0
    batch_size: 4096  Epoch  22 mean_squared_error: 6.02740e-05 val_mean_squared_error: 1.52414e-04

Based on what I'm seeing, I might think that the beta values are too low....(or soul I rather say they are decreasing to fast?). Maybe I should reset the values in-between each epoch?

**Comment by oysteijo** [comment]
This is really soon within my goal, and I can close this..... I think I will have to use the OpenMP (not commited code.) enabled code on the NC24 node in Azure. I then get one epoch each 30 seconds.

**Comment by oysteijo** [comment]
I'm actually quite solid under 1.5e-4 using AdamW. I think the best is 1.467e-4 (?). I can maybe get it even lower? It looks like adding weight decay actually helps a lot. :1st_place_medal: 

Edit: The final number is:

    [11:39:41] Epoch 339 mean_squared_error: 5.66518e-05 val_mean_squared_error: 1.46797e-04

.... so it is 1.46**8**e-4.
